### PR TITLE
operator: release leader lease lock on operator exit

### LINKED ISSUE #14546: CIlium operator HA lease lock is not released properly
When running `cilium-operator` in HA mode if the operator instance acting as leader exits on the interception of a signal the lease is not released properly. This happens because we preemptively do an `os.Exit()` before the leader election library can release the lock it holds.

https://github.com/cilium/cilium/blob/f2163862a42494030fd4177ddf25f021c85ae2e3/operator/main.go#L151-L152

Here when the `leaderElectionContext` is canceled we don't wait for the lease to be released by the library and exit preemptively. This leads to a situation where the lease has a leader who does not exist. For the other operator instance to own this lease we have to wait for the lease to expire. So essentially there is a delay for the other operator to become the leader. 

This can be illustrated more clearly when we have a single operator instance. If the operator process is restarted the new process will report that the leader exists with the id of the previous instance. Then only after the lease expires the operator will start leading.

Waiting on #14484 to be merged before fixing this as the current operator API server implementation does not have a proper shutdown implementation in place.

### PR DESCRIPTION
**Pull request by fristonio**
Fixes #14546 

The operator on receiving the termination signal will wait for the lease to be released and then exit. 

```
Jan 07 18:42:21 k8s1 cilium-operator[4918]: level=debug msg="Controller func execution time: 1.809µs" name=update-cilium-nodes-pod-cidr subsys=controller uuid=cc0277ad-5270-4335-b8ce-eaacc95e1717
Jan 07 18:42:23 k8s1 systemd[1]: Stopping cilium operator...
Jan 07 18:42:23 k8s1 cilium-operator[4918]: level=info msg="Received termination signal. Shutting down" subsys=cilium-operator
Jan 07 18:42:23 k8s1 cilium-operator[4918]: level=info msg="Leader election lost" operator-id=k8s1-jUBZGmHGKZ subsys=cilium-operator
Jan 07 18:42:23 k8s1 systemd[1]: Stopped cilium operator.
```

### CODE REVIEW

File: operator/api/server.go [hunk 1]
```diff
@@ -152,16 +160,21 @@ func (s *Server) StartServer() error {
...
 			case err := <-errCh:
-				log.WithError(err).Warn("Unable to start status api")
+				log.WithError(err).Warn("Unable to start operator API server")
 			}
 		}()
+
 		log.Infof("Starting apiserver on address %s", addr)
 	}
 
 	for err := range errs {
```
Comment (christarazi) [inline operator/api/server.go:171]: I think we can move L105 to just above this for-loop so that the variable is closer in scope to its use.

**Review by christarazi — APPROVED** [review]
:rocket:

File: operator/api/server.go [hunk 2]
```diff
@@ -140,8 +141,14 @@ func (s *Server) StartServer() error {
 		go func() {
 			err := srv.Serve(ln)
 			if err != nil {
-				errCh <- err
-				errs <- err
+				// If the error is due to the server being shutdown, then send nil to
+				// the server errors channel.
+				if errors.Is(err, http.ErrServerClosed) {
+					errs <- nil
```
Comment (aanm) [inline operator/api/server.go:148]: might be worth printing out a message stating "server closed" or something like that.

**Review by aanm — APPROVED** [review]
Small non-blocking nit.

**Comment by fristonio** [comment]
test-me-please

**Comment by fristonio** [comment]
retest-gke

**Comment by joestringer** [comment]
Looks like this fell through the cracks in 1.9 backports.

@errordeveloper reported in the [linked PR](https://github.com/cilium/cilium/pull/14613):
> @fristonio as you suggested, I will drop #14554 as it depends on a larger PR #14484.

Opting to drop the 1.9 backport label. @fristonio , if you think this is important to backport to v1.9 then please either initiate the backport or reach out in #launchpad on slack to coordinate.

**Pull request merged**